File: android/app/src/main/java/es/manabe/yomiyasu/core/settings/AppSettings.kt
package es.manabe.yomiyasu.core.settings

import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.booleanPreferencesKey
import androidx.datastore.preferences.core.doublePreferencesKey
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.intPreferencesKey
import androidx.datastore.preferences.core.stringPreferencesKey
import es.manabe.yomiyasu.app.ui.theme.ThemeMode
import es.manabe.yomiyasu.core.di.ApplicationScope
import es.manabe.yomiyasu.core.models.MainView
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import javax.inject.Inject
import javax.inject.Singleton

enum class BookViewMode(val rawValue: String) {
    Characters("characters"),
    Pages("pages"),
    Both("both"),
    RemainingCharacters("remainingCharacters"),
    RemainingPages("remainingPages"),
    RemainingTime("remainingTime"),
    ;

    val title: String
        get() = when (this) {
            Characters -> "Caracteres"
            Pages -> "Páginas"
            Both -> "Ambos"
            RemainingCharacters -> "Caracteres restantes"
            RemainingPages -> "Páginas restantes"
            RemainingTime -> "Tiempo restante"
        }
}

data class BoardVisibility(
    val progress: Boolean = true,
    val tablero: Boolean = true,
    val readLater: Boolean = true,
    val paused: Boolean = false,
    val newBooks: Boolean = true,
    val newSeries: Boolean = true,
    val recentSeries: Boolean = true,
)

enum class BoardFlag {
    Progress,
    Tablero,
    ReadLater,
    Paused,
    NewBooks,
    NewSeries,
    RecentSeries,
}

data class AppSettingsData(
    val mainView: MainView = MainView.Both,
    val antispoilers: Boolean = false,
    val appearance: ThemeMode = ThemeMode.Dark,
    val autoCrono: Boolean = false,
    val showCrono: Boolean = true,
    val bookView: BookViewMode = BookViewMode.Characters,
    val meanCharactersPerHour: Double? = null,
    val idleTimeout: Int = 0,
    val boards: BoardVisibility = BoardVisibility(),
)

@Singleton
class AppSettings @Inject constructor(
    private val dataStore: DataStore<Preferences>,
    @ApplicationScope private val scope: CoroutineScope,
) {
    private object Key {
        val MainView = stringPreferencesKey("mainView")
        val Antispoilers = booleanPreferencesKey("antispoilers")
        val Appearance = stringPreferencesKey("appearance")
        val AutoCrono = booleanPreferencesKey("autoCrono")
        val ShowCrono = booleanPreferencesKey("showCrono")
        val BookView = stringPreferencesKey("bookView")
        val MeanSpeed = doublePreferencesKey("meanSpeed")
        val IdleTimeout = intPreferencesKey("idleTimeout")
        val BoardProgress = booleanPreferencesKey("boardProgress")
        val BoardTablero = booleanPreferencesKey("boardTablero")
        val BoardReadLater = booleanPreferencesKey("boardReadLater")
        val BoardPaused = booleanPreferencesKey("boardPaused")
        val BoardNewBooks = booleanPreferencesKey("boardNewBooks")
        val BoardNewSeries = booleanPreferencesKey("boardNewSeries")
        val BoardRecentSeries = booleanPreferencesKey("boardRecentSeries")
    }

    private fun boardKey(flag: BoardFlag): Preferences.Key<Boolean> = when (flag) {
        BoardFlag.Progress -> Key.BoardProgress
        BoardFlag.Tablero -> Key.BoardTablero
        BoardFlag.ReadLater -> Key.BoardReadLater
        BoardFlag.Paused -> Key.BoardPaused
        BoardFlag.NewBooks -> Key.BoardNewBooks
        BoardFlag.NewSeries -> Key.BoardNewSeries
        BoardFlag.RecentSeries -> Key.BoardRecentSeries
    }

    private fun boardsOf(prefs: Preferences): BoardVisibility = BoardVisibility(
        progress = prefs[Key.BoardProgress] ?: true,
        tablero = prefs[Key.BoardTablero] ?: true,
        readLater = prefs[Key.BoardReadLater] ?: true,
        paused = prefs[Key.BoardPaused] ?: false,
        newBooks = prefs[Key.BoardNewBooks] ?: true,
        newSeries = prefs[Key.BoardNewSeries] ?: true,
        recentSeries = prefs[Key.BoardRecentSeries] ?: true,
    )

    val flow: StateFlow<AppSettingsData> = dataStore.data
        .map { prefs ->
            AppSettingsData(
                mainView = prefs[Key.MainView]
                    ?.let { value -> MainView.entries.firstOrNull { it.rawValue == value } }
                    ?: MainView.Both,
                antispoilers = prefs[Key.Antispoilers] ?: false,
                appearance = prefs[Key.Appearance]
                    ?.let { value -> ThemeMode.entries.firstOrNull { it.name.lowercase() == value } }
                    ?: ThemeMode.Dark,
                autoCrono = prefs[Key.AutoCrono] ?: false,
                showCrono = prefs[Key.ShowCrono] ?: true,
                bookView = prefs[Key.BookView]
                    ?.let { value -> BookViewMode.entries.firstOrNull { it.rawValue == value } }
                    ?: BookViewMode.Characters,
                meanCharactersPerHour = prefs[Key.MeanSpeed],
                idleTimeout = prefs[Key.IdleTimeout] ?: 0,
                boards = boardsOf(prefs),
            )
        }
        .stateIn(scope, SharingStarted.Eagerly, AppSettingsData())

    suspend fun current(): AppSettingsData = flow.first()

    suspend fun setMainView(value: MainView) {
        dataStore.edit { it[Key.MainView] = value.rawValue }
    }

    suspend fun setAntispoilers(value: Boolean) {
        dataStore.edit { it[Key.Antispoilers] = value }
    }

    suspend fun setAppearance(value: ThemeMode) {
        dataStore.edit { it[Key.Appearance] = value.name.lowercase() }
    }

    suspend fun setAutoCrono(value: Boolean) {
        dataStore.edit { it[Key.AutoCrono] = value }
    }

    suspend fun setShowCrono(value: Boolean) {
        dataStore.edit { it[Key.ShowCrono] = value }
    }

    suspend fun setBookView(value: BookViewMode) {
        dataStore.edit { it[Key.BookView] = value.rawValue }
    }

    suspend fun setMeanCharactersPerHour(value: Double?) {
        dataStore.edit { prefs ->
            if (value != null) {
                prefs[Key.MeanSpeed] = value
            } else {
                prefs.remove(Key.MeanSpeed)
            }
        }
    }

    suspend fun setBoard(flag: BoardFlag, value: Boolean) {
        dataStore.edit { prefs -> prefs[boardKey(flag)] = value }
    }

    suspend fun setIdleTimeout(value: Int) {
        dataStore.edit { it[Key.IdleTimeout] = value }
    }
}

File: ios/Yomiyasu/Core/Settings/AppSettings.swift
import Foundation
import Observation
import SwiftUI

enum AppAppearance: String, CaseIterable, Identifiable, Codable {
    case system
    case light
    case dark

    var id: String { rawValue }

    var title: String {
        switch self {
        case .system: "Sistema"
        case .light: "Claro"
        case .dark: "Oscuro"
        }
    }

    var colorScheme: ColorScheme? {
        switch self {
        case .system: nil
        case .light: .light
        case .dark: .dark
        }
    }
}

enum BookViewMode: String, CaseIterable, Identifiable, Codable {
    case characters
    case pages
    case both
    case remainingCharacters
    case remainingPages
    case remainingTime

    var id: String { rawValue }

    var title: String {
        switch self {
        case .characters: "Caracteres"
        case .pages: "Páginas"
        case .both: "Ambos"
        case .remainingCharacters: "Caracteres restantes"
        case .remainingPages: "Páginas restantes"
        case .remainingTime: "Tiempo restante"
        }
    }
}

struct BoardVisibility: Equatable, Sendable {
    var progress = true
    var tablero = true
    var readLater = true
    var paused = false
    var newBooks = true
    var newSeries = true
    var recentSeries = true
}

@MainActor
@Observable
final class AppSettings {
    private enum Key {
        static let mainView = "mainView"
        static let antispoilers = "antispoilers"
        static let appearance = "appearance"
        static let autoCrono = "autoCrono"
        static let showCrono = "showCrono"
        static let bookView = "bookView"
        static let meanSpeed = "meanSpeed"
        static let idleTimeout = "idleTimeout"
        static let boardProgress = "boardProgress"
        static let boardTablero = "boardTablero"
        static let boardReadLater = "boardReadLater"
        static let boardPaused = "boardPaused"
        static let boardNewBooks = "boardNewBooks"
        static let boardNewSeries = "boardNewSeries"
        static let boardRecentSeries = "boardRecentSeries"
    }

    private let defaults: UserDefaults

    var mainView: MainView {
        didSet { defaults.set(mainView.rawValue, forKey: Key.mainView) }
    }

    var antispoilers: Bool {
        didSet { defaults.set(antispoilers, forKey: Key.antispoilers) }
    }

    var appearance: AppAppearance {
        didSet { defaults.set(appearance.rawValue, forKey: Key.appearance) }
    }

    var autoCrono: Bool {
        didSet { defaults.set(autoCrono, forKey: Key.autoCrono) }
    }

    var showCrono: Bool {
        didSet { defaults.set(showCrono, forKey: Key.showCrono) }
    }

    var bookView: BookViewMode {
        didSet { defaults.set(bookView.rawValue, forKey: Key.bookView) }
    }

    var meanCharactersPerHour: Double? {
        didSet {
            if let meanCharactersPerHour {
                defaults.set(meanCharactersPerHour, forKey: Key.meanSpeed)
            } else {
                defaults.removeObject(forKey: Key.meanSpeed)
            }
        }
    }

    var boards: BoardVisibility {
        didSet { persistBoards() }
    }

    /// Minutos sin actividad de lectura tras los que se pausa el cronómetro. 0 lo desactiva.
    var idleTimeout: Int {
        didSet { defaults.set(idleTimeout, forKey: Key.idleTimeout) }
    }

    private func persistBoards() {
        defaults.set(boards.progress, forKey: Key.boardProgress)
        defaults.set(boards.tablero, forKey: Key.boardTablero)
        defaults.set(boards.readLater, forKey: Key.boardReadLater)
        defaults.set(boards.paused, forKey: Key.boardPaused)
        defaults.set(boards.newBooks, forKey: Key.boardNewBooks)
        defaults.set(boards.newSeries, forKey: Key.boardNewSeries)
        defaults.set(boards.recentSeries, forKey: Key.boardRecentSeries)
    }

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        mainView = MainView(rawValue: defaults.string(forKey: Key.mainView) ?? "") ?? .both
        antispoilers = defaults.bool(forKey: Key.antispoilers)
        appearance = AppAppearance(
            rawValue: defaults.string(forKey: Key.appearance) ?? ""
        ) ?? .dark
        autoCrono = defaults.bool(forKey: Key.autoCrono)
        showCrono = defaults.object(forKey: Key.showCrono) as? Bool ?? true
        bookView = BookViewMode(
            rawValue: defaults.string(forKey: Key.bookView) ?? ""
        ) ?? .characters
        meanCharactersPerHour = defaults.object(forKey: Key.meanSpeed) as? Double
        idleTimeout = defaults.object(forKey: Key.idleTimeout) as? Int ?? 0
        boards = BoardVisibility(
            progress: defaults.object(forKey: Key.boardProgress) as? Bool ?? true,
            tablero: defaults.object(forKey: Key.boardTablero) as? Bool ?? true,
            readLater: defaults.object(forKey: Key.boardReadLater) as? Bool ?? true,
            paused: defaults.object(forKey: Key.boardPaused) as? Bool ?? false,
            newBooks: defaults.object(forKey: Key.boardNewBooks) as? Bool ?? true,
            newSeries: defaults.object(forKey: Key.boardNewSeries) as? Bool ?? true,
            recentSeries: defaults.object(forKey: Key.boardRecentSeries) as? Bool ?? true
        )
    }
}
